src/box/decimal.h

```diff
@@ -0,0 +1,586 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2022, Tarantool AUTHORS, please see AUTHORS file.
+ */
+#pragma once
+
+#include "trivia/util.h"
+
+#if defined(__cplusplus)
+extern "C" {
+#endif /* defined(__cplusplus) */
+
+/** \cond public */
+
+/* {{{ decimal structure and constants */
+
+/** See @ref box_decimal_storage_t. */
+struct box_decimal_storage {
+	/**
+	 * Decimal data.
+	 *
+	 * The format is unspecified and can be changed.
+	 */
+	uint64_t internal[8];
+};
+
+/**
+ * Storage for a decimal number.
+ *
+ * This type is large enough to hold a decimal value. The primary
+ * usage is to allocate a decimal on the stack and pass it to a
+ * box_decimal_*() function.
+ *
+ * Usage example:
+ *
+ *  | box_decimal_storage_t storage;
+ *  | box_decimal_t *dec = (box_decimal_t *)&storage;
+ *  | box_decimal_<...>(dec);
+ *
+ * The only legal usage is to allocate a memory for the structure
+ * and cast its address to (box_decimal_t *) to use with
+ * box_decimal_*() functions.
+ */
+typedef struct box_decimal_storage box_decimal_storage_t;
```

Comment:
I think that a decimal number will be more often allocated as a part of a struct or array or on stack, not with `malloc`. Having a separate storage class will complicate these use cases so I lean towards [v2](https://github.com/tarantool/tarantool/pull/7497).

Reply:
My doubts are around [luaT_newdecimal](https://github.com/tarantool/tarantool/pull/7429)(): here we get `box_decimal_t *` pointer and must use `box_decimal_copy()` (or `box_decimal_from_*()`) to fill it. However it is quite easy to make a mistake and do something like this:

```c
box_decimal_t *res = luaT_newdecimal(L);
*res = *dec; /* !! out-of-bound access !! */
return 1;
```

It'll write out of the decimal structure.

I can't compare those 'often X, not Y' (how often?), 'more complicated' (how much?), 'easy to make a mistake' (how easy?).

If your opinion was not changed by the example above, I'll close this PR and we'll go in the v2 direction.

Reply:
Maybe, push 64 bytes to Lua stack then?

```diff
diff --git a/src/lua/decimal.c b/src/lua/decimal.c
index a50e4142a2fe..53587c524174 100644
--- a/src/lua/decimal.c
+++ b/src/lua/decimal.c
@@ -427,10 +427,7 @@ void
 tarantool_lua_decimal_init(struct lua_State *L)
 {
 	int rc = luaL_cdef(L, "typedef struct {"
-				       "int32_t digits;"
-				       "int32_t exponent;"
-				       "uint8_t bits;"
-				       "uint16_t lsu[13];"
+				       "uint64_t internal[8];"
 			      "} decimal_t;");
 	assert(rc == 0);
 	(void)rc;
```

Reply:
v2 seems simpler in my opinion. @locker's suggestion with pushing 64 bytes to Lua stack seems scary, at first, because we're consuming almost twice the amount of memory we actually use.
But I'm okay with this as long as storing decimals doesn't require this extra padding.

After giving it a second thought, maybe we enlarge decimal storage to ~ 70-ish digits (to store (u)int256) right now?
Then there will be no regrets about the lost space.
Besides, all there is to change is one define.

Reply:
> Maybe, push 64 bytes to Lua stack then?

What doubts me here:

1. Lua memory is a tightly bounded resource: AFAIR, 2 or 4 GiB at max in non-gc-64 LuaJIT build (which is the primary build type on Linux). If an application works with decimals intensively, will not we make the limit more annoying than before for administrators?
2. It is unclear, whether we can declare `*res = *dec` as legal assignment and drop `decimal_copy()`. Is it possible that some other API will return a pointer to a decimal value allocated within tarantool (which have less than 64 bytes allocated)?